File: api-laravel/Modules/Place/app/Services/Geocoder/Geocoder.php
<?php

namespace Modules\Place\Services\Geocoder;

interface Geocoder
{
    /**
     * @param  array{street?: ?string, city?: ?string, province?: ?string, postal_code?: ?string, country?: ?string}  $address
     * @return array{latitude: ?float, longitude: ?float}
     */
    public function geocode(array $address): array;
}

File: api-go/pkg/geocoder/geocoder.go
package geocoder

import "context"

// Address holds fields used for geocoding.
type Address struct {
	Street     string
	City       string
	Province   string
	PostalCode string
	Country    string
	Latitude   *float64
	Longitude  *float64
}

// Coordinates is a geocoding result.
type Coordinates struct {
	Latitude  *float64
	Longitude *float64
}

// Geocoder resolves coordinates for an address.
type Geocoder interface {
	Geocode(ctx context.Context, addr Address) (Coordinates, error)
}

// Noop returns manual coordinates when provided.
type Noop struct{}

func (Noop) Geocode(_ context.Context, addr Address) (Coordinates, error) {
	return Coordinates{Latitude: addr.Latitude, Longitude: addr.Longitude}, nil
}
